### ISSUE TITLE
SelectQueryImpl.asTable() alias generation is slow

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#6714]
The automatic alias generation in `SelectQueryImpl.asTable()` uses `Tools.hash()` internally, which relies on generating the SQL string for the query. This is very slow and should be improved.

Currently, it is unclear how a sufficiently unique alias can be generated this way. Perhaps we should consider generating aliases from the context of where a derived table is embedded, making the alias non-reproduceable (or even non-existent) in standalone calls.

Note, this does not affect `SelectQueryImpl.asTable(String)`, where an explicit alias is provided.

**Comment by lukaseder** [comment]
If `SelectQueryImpl` was immutable, we could at least cache the `asTable()` call. Currently, we cannot.

**Comment by lukaseder** [comment]
This problem appears time and again in different contexts, not just when users call the method. Users can work around the problem easily by providing an explicit alias, but jOOQ's internals often cannot, as the call is generated indirectly, e.g. when calling `Fields::fields`:

![image](https://user-images.githubusercontent.com/734593/160352783-f49d861a-2c6a-43b4-a58e-38f387baa441.png)

A new integration test case could help prevent such calls from jOOQ internals: https://github.com/jOOQ/jOOQ/issues/13341

**Comment by juriad** [comment]
In our company we go a step further - every call of `asTable()` without parameter is wrong - our tables are very wide and this hurts us a lot (performance-wise).

We would welcome an option to specify an alias generator. Or go even a step further and deprecate the method with no parameter - if jOOQ internally stops calling it, why should the user? Or completely remove interface `TableLike` - because the developers too often don't pass Selects where Tables are logically expected (from, join).

We were going through the codebase and were adding the explicit aliases until we "fixed" it by the following code. It occasionally breaks during upgrades though but generally works well with the way we use jOOQ.

```
@UtilityClass
class FixJooqSlowAliasing {
    /**
     * It fixes slow alias rendering in jooq Tools class. Tools uses CTX static field of DSLContext type
     * to render QueryParty and then creates hashCode of the rendered String. And it only uses
     * render method from the CTX field. This sets CTX to our own DSLContext object which implements
     * only render method which renders the QueryPart to identity hash code. Which is much faster
     * but it does not render sql and it would be a problem if it was used anywhere else than here.
     */
    @SneakyThrows
    void fix() {
        Class<?> toolsClass = DSLContext.class.getClassLoader().loadClass("org.jooq.impl.Tools");
        Field field = toolsClass.getDeclaredField("CTX");
        setNonFinal(field);
        field.setAccessible(true);
        field.set(null, CTX);
    }

    private void setNonFinal(Field field) throws NoSuchFieldException, IllegalAccessException {
        Field modifiersField = Field.class.getDeclaredField("modifiers");
        modifiersField.setAccessible(true);
        modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
    }

    private static final DSLContext CTX = createCtx();

    @SneakyThrows
    private static DSLContext createCtx() {
        DefaultConfiguration configuration = new DefaultConfiguration();

        return (DSLContext) Proxy.newProxyInstance(DSLContext.class.getClassLoader(), new Class[] {DSLContext.class}, (proxy, method, args) -> {
            if (method.getName().equals("render") && Arrays.equals(method.getParameterTypes(), new Class[] {QueryPart.class})) {
                return Integer.toString(System.identityHashCode(args[0]));
            }
            if (method.getName().equals("configuration") && method.getParameterTypes().length == 0) {
                return configuration;
            }

            throw new IllegalStateException("Internal error in workaround code in jooq library. Method " + method + " was called.");
        });
    }
}
```

**Comment by lukaseder** [comment]
Thanks for your feedback, @juriad, greatly appreciated.

> In our company we go a step further - every call of `asTable()` without parameter is wrong - our tables are very wide and this hurts us a lot (performance-wise).

I'm curious about benchmarks / profiling sessions. How does this "hurting" manifest? Do you have data you can share?

> We would welcome an option to specify an alias generator. Or go even a step further and deprecate the method with no parameter - if jOOQ internally stops calling it, why should the user? Or completely remove interface `TableLike` - because the developers too often don't pass Selects where Tables are logically expected (from, join).

It's easy to get lured into exaggeration, but let's try not to.

1. jOOQ internally calling it causes trouble that the user cannot prevent, so this is the significant part I'm currently addressing in https://github.com/jOOQ/jOOQ/issues/13341. There's no workaround for this, except for hacks like yours.
2. The user calling it doesn't *have to* be a problem. While the 2.4% of the profiling session shown here https://github.com/jOOQ/jOOQ/issues/6714#issuecomment-1080326811 are significant enough to address the issue (2.4% of the total execution time of some `MULTISET` query, which includes the execution time on the server), it's not the end of the world. E.g. we're not spending 50% on this logic.
3. In many cases, aliases are completely unnecessary. I don't like the fact that many RDBMS require them *everywhere*. For example, if you have a derived table and just want to project everything from it, producing e.g. a `Record3<Integer, String, Date>`, why should you give it a name?
4. Some RDBMS allow for omitting the table alias under some circumstances. This includes Oracle. We need to be able to handle this kind of SQL in the parser and elsewhere *somehow*, so why not just support this in the jOOQ API natively?
5. The only reason why this hasn't been fixed yet, is because there's a significant risk in changing such a central piece of logic thoroughly. That doesn't mean it can't be done, nor that we have to change tons of other things that are not strictly related, including removing `TableLike`. The issue is old and there are zero upvotes on it, so the takeaway is that this can't be a *very* important problem for most users.

So, let's not escalate this bug into a design flaw, which it isn't. It's just a bug.

Now, even if this issue here were fixed, https://github.com/jOOQ/jOOQ/issues/13341 would be a reasonable thing to do, because even if the generated alias is generated much more quickly, it would still be slower than a *constant* alias that is used whenever the alias is irrelevant. So, #13341 is a quick fix with little risk and significant benefit. This issue here #6714 is a tricky fix with lots of risk and much less benefit.

> We were going through the codebase and were adding the explicit aliases until we "fixed" it by the following code. It occasionally breaks during upgrades though but generally works well with the way we use jOOQ.

Your solution works for you, but be very careful with that identity hash code thing. A *much more* significant flaw that your fix would produce on any commercial RDBMS is that such queries would now produce unstable SQL strings, which would cause *a lot* of contention on the execution plan cache. This may not be a problem in PostgreSQL, but it would be a *huge* problem e.g. in Oracle, and the very reason why we generate the alias based on a stable hash code like we currently do.

Also, when you use the identity hash code, this might prevent `QueryPart::equals` and `QueryPart::hashCode` from working correctly elsewhere, whose default behaviour is also based on `toString()`, which is stable. Not sure if this is relevant to you, but certainly good to know.

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/10277 has introduced (among other things) a `LazyName` object, which contains:

- A name computation
- A transient, cached name

The class is here: https://github.com/jOOQ/jOOQ/blob/main/jOOQ/src/main/java/org/jooq/impl/LazyName.java

It is a placeholder that can be used to delay the calculation of a name. It could be useful in this case here, where a lot of alias generations are incidental and wouldn't be needed, e.g. if `Stream.of(Select.asTable().fields()).map(f -> f.getDataType())` or other similar things are being done. One such example is this idiom here:

```java
// This
table(select).as("x")

// is equivalent to this:
select.asTable("x")
```

See also https://github.com/jOOQ/jOOQ/issues/13349. In the above case, currently, the intermediate alias is computed slowly without ever being used, as it is immediately overridden by an explicit alias.

**Comment by lukaseder** [comment]
The accidental practice of using `select.asTable().as("x")` or `table(select).as("x")` instead of `select.asTable("x")` has now been improved. The `asTable()` call sufficiently delays the alias computation until it is really necessary, which it isn't in the previous examples where an explicit alias is supplied right after.

The relevant improvement is: https://github.com/jOOQ/jOOQ/issues/13349 and it has been implemented in 3.17.0, 3.16.6, and 3.15.10.

**Comment by lukaseder** [comment]
> We would welcome an option to specify an alias generator

I noticed I haven't replied to this suggestion yet. There's a pending feature request, which does something similar: Offering an SPI that overrides top level projection aliases: https://github.com/jOOQ/jOOQ/issues/11545

Your suggestion is a bit different. It would allow for overriding the algorithm of `org.jooq.impl.Tools::autoAlias`. Both of these SPIs seem occasionally useful, but currently don't pull their weight. Specifically, in the latter case, any implementation suggestion that is viable might as well be implemented natively by jOOQ.

**Comment by juriad** [comment]
Thanks for all the improvements. It is very well possible that the problem is gone or significantly smaller than it used to be 2 years ago. I will try to disable our "fix" and profile the application but it will take me time.

Meanwhile, I prepared a synthetic benchmark that generates a query similar to:
```
SELECT sub.* - rn
FROM (SELECT *, row_number() over () as rn FROM table_with_1000_columns) sub
WHERE rn > 1
```
where the returned columns are all columns without rn qualified by subquery alias. The pattern that we often have is:
* a wide table (100 columns are common; 1000 not, but look better in benchmark)
* a subselect that cannot be inlined (due to window function, group by, pagination)
* the list of columns returned comes mainly from the subselect and they must be aliases because there are other joins on the outer level that have conflicting names

https://gist.github.com/juriad/91db1830b78fb89b992c42bf4e3a4348

It is interesting to see that explicitly setting alias to the subselect (calling `asTable("sub")` helps a lot. 
An even better solution is not to call `TableLike.field()` but to construct the fields manually. 

`TableLike.field()` is quadratic in time when called for all fields of `Table` and even worse when called on `Select`. (A solution might be to use indices instead of names or fields themselves but the experience suffers.)

How can we make the fastest option easier to write? Am I missing any API?

**Comment by lukaseder** [comment]
> Thanks for all the improvements. It is very well possible that the problem is gone or significantly smaller than it used to be 2 years ago

Well, this is the first time this stuff has been improved in a while, so before the fixes are released, I don't think you'll see much improvement (unless you avoid calling `asTable()` from your end, of course)

> I will try to disable our "fix" and profile the application but it will take me time.

I appreciate it. That will help find other cases that I may have overlooked so far.

> Meanwhile, I prepared a synthetic benchmark that generates a query similar to [...]

Before we move on, is this strictly related to `asTable()` alias generation? `TableLike.field()` might be, because we're also generating aliases in that method (given that the field needs a proper table qualification).

But perhaps, that would be discussing a side-effect of this issue here? I'd rather separate concerns, so issues can be tracked separately and possibly solved separately. The fact that some lookup is quadratic probably isn't related to `SelectQueryImpl.asTable()`

**Comment by lukaseder** [comment]
... Regarding the quadratic algorithm, I mean, yes it's possible to fall into this trap by accident, and yes, this issue #6714 here makes things very much worse. But this stream here:

```java
TABLE.fieldStream().map(sub::field).collect(Collectors.toList());
```

Is kinda akin to:

```
// This translation is also quadradic
list.map(e -> list.get(list.indexOf(e)))
```

Since you already know what you did when you constructed the query, it seems you could just do this, instead (on the derived table with explicit alias `"sub"`, of course)?

```
sub.fieldStream().limit(TABLE.fields().length).toList();
```

**Comment by lukaseder** [comment]
As far as quadradic lookups are concerned, I've thought about an alternative internal implementation of the `org.jooq.impl.FieldsImpl` class, which models all the `Field` lookups. For large degrees (e.g. `> 32`, or whatever), it might be worth thinking about storing things in a `HashMap` instead of a `Field<?>[]`, which is a more optimal data structure for smaller degrees.

That change would effectively turn a quadratic algorithm to a linear one.

I've created: https://github.com/jOOQ/jOOQ/issues/13384

**Comment by lukaseder** [comment]
> The relevant improvement is: #13349 and it has been implemented in 3.17.0, 3.16.6, and 3.15.10.

I think that this fix is sufficient. I'll close this as a duplicate of #13349
